src/Engine/IntegrityCheckTrait.php

```diff
@@ -98,31 +98,24 @@ protected function doCheckIntegrity(CoreContextAwarePluginInterface $plugin, Exe
       $provided_context_definitions = $plugin->getProvidedContextDefinitions();
 
       foreach ($provided_context_definitions as $name => $context_definition) {
-        if (isset($this->configuration['provides_mapping'][$name])) {
-          if (!preg_match('/^[0-9a-zA-Z_]*$/', $this->configuration['provides_mapping'][$name])) {
-            $violation = new IntegrityViolation();
-            $violation->setMessage($this->t('Provided variable name %name contains not allowed characters.', [
-              '%name' => $this->configuration['provides_mapping'][$name],
-            ]));
-            $violation->setContextName($name);
-            $violation->setUuid($this->getUuid());
-            $violation_list->add($violation);
-          }
-
-          // Populate the state with the new variable that is provided by this
-          // plugin. That is necessary so that the integrity check in subsequent
-          // actions knows about the variable and does not throw violations.
-          $metadata_state->setDataDefinition(
-            $this->configuration['provides_mapping'][$name],
-            $context_definition->getDataDefinition()
-          );
-        }
-        else {
-          $metadata_state->setDataDefinition($name, $context_definition->getDataDefinition());
+        if (isset($this->configuration['provides_mapping'][$name])
+          && !preg_match('/^[0-9a-zA-Z_]*$/', $this->configuration['provides_mapping'][$name])
+        ) {
+          $violation = new IntegrityViolation();
+          $violation->setMessage($this->t('Provided variable name %name contains not allowed characters.', [
+            '%name' => $this->configuration['provides_mapping'][$name],
+          ]));
+          $violation->setContextName($name);
+          $violation->setUuid($this->getUuid());
+          $violation_list->add($violation);
         }
       }
     }
 
+    // Make sure that all proived variables by this plugin are added to the
+    // execution metadata state.
+    $this->doPrepareExecutionMetadataState($plugin, $metadata_state);
```

Comment:
oh I see now that the plugin is required to update that method with suiting state adaptions as necessary - however that's not obvious for the caller / wasn't obvious for me :) Thus this needs to be documented.
Besides that, I wonder whether that would not belong into ContextHandlerTrait really? The trait cares already of the context mapping for execution time, so would it make sense to care about at configuration time as well? Or if we want to separate it for better code organization, we could make a ExecutionContextHandlerTrait & a ExecutionMetadataHandlerTrait below the Context namespace?